File: python3-sdk/semantria/authrequest.py
# -*- coding: utf-8 -*-

import http.client
import time
import random
import urllib.parse
import hmac
import binascii
import hashlib

OAuthVersion = "1.0"
OAuthParameterPrefix = "oauth_"
OAuthConsumerKeyKey = "oauth_consumer_key"
OAuthVersionKey = "oauth_version"
OAuthSignatureMethodKey = "oauth_signature_method"
OAuthSignatureKey = "oauth_signature"
OAuthTimestampKey = "oauth_timestamp"
OAuthNonceKey = "oauth_nonce"

class AuthRequest:
    def __init__(self, consumerKey, consumerSecret, applicationName):
        self.consumerKey = consumerKey
        self.consumerSecret = consumerSecret
        self.applicationName = applicationName

    def authWebRequest(self, method, url, postData):
        nonce = self.generateNonce()
        timestamp = self.generateTimestamp()

        query = self.generateQuery(method, url, timestamp, nonce)
        authheader = self.generateAuthHeader(query, timestamp, nonce)
        
        headers = {"Authorization": authheader}
        if (method == "POST"):
            headers["Content-type"] = "application/x-www-form-urlencoded"
        headers["x-api-version"] = "2"
        headers["x-app-name"] = self.applicationName
        
        qparts = urllib.parse.urlparse(query)
        qscheme, qnetloc, qpath, qparams, qquery, qfragment = qparts[:6]
        conn = http.client.HTTPSConnection(qnetloc)
        host = '%s?%s' % (qpath, qquery)
        conn.request(method, host, postData, headers)
        response = conn.getresponse()
        result = { "status":response.status, "reason":response.reason, "data":response.read() }
        conn.close()
        return result

    def generateQuery(self, method, url, timestamp, nonce):
        parts = urllib.parse.urlparse(url)
        scheme, netloc, path, params, query, fragment = parts[:6]
        np = self.getNormalizedParameters(timestamp, nonce)

        if query:
            query = '%s&oauth_verifier=%s' % (query, np)
        else:
            query = '%s' % np
        
        return urllib.parse.urlunparse((scheme, netloc, path, params, query, fragment))         

    def generateAuthHeader(self, query, timestamp, nonce):
        md5cs = self.getMD5Hash(self.consumerSecret)
        escquery = self.escape(query)
        hash = self.getSHA1(md5cs, escquery)
        hash = str(hash, encoding = 'utf-8')
        hash = self.escape(hash)
       
        items = []
        items.append(("OAuth realm", ""))
        items.append((OAuthVersionKey, "%s" % OAuthVersion))
        items.append((OAuthTimestampKey, "%s" % timestamp))
        items.append((OAuthNonceKey, "%s" % nonce))
        items.append((OAuthSignatureMethodKey, "HMAC-SHA1"))
        items.append((OAuthConsumerKeyKey, "%s" % self.consumerKey))
        items.append((OAuthSignatureKey, "%s" % hash))
        items.sort()

        params = []
        for k, v in items:
            params.append('%s="%s"' % (k, v))

        return ",".join(params)

    def getNormalizedParameters(self, timestamp, nonce):
        items = []
        items.append((OAuthVersionKey, OAuthVersion))
        items.append((OAuthTimestampKey, timestamp))
        items.append((OAuthNonceKey, nonce))
        items.append((OAuthSignatureMethodKey, "HMAC-SHA1"))
        items.append((OAuthConsumerKeyKey, self.consumerKey))
        items.sort()
        
        encoded_str = urllib.parse.urlencode(items)
        # Encode signature parameters per Oauth Core 1.0 protocol
        # Spaces must be encoded with "%20" instead of "+"
        return encoded_str.replace('+', '%20').replace('%7E', '~')

    def getMD5Hash(self, str):
        md5h = hashlib.md5()
        md5h.update(str.encode('utf-8'))
        md5sig = md5h.hexdigest()
        return md5sig
 
    def getSHA1(self, md5cs, query):
        sha1res = hmac.new(bytes(md5cs.encode('utf-8')), bytes(query.encode('utf-8')), hashlib.sha1).digest()
        sha1sig = binascii.b2a_base64(sha1res)[:-1]
        return sha1sig

    def generateTimestamp(self):
        return int(time.time())

    def generateNonce(self, length=20):
        return ''.join([str(random.randint(0, 9)) for i in range(length)])
    
    def escape(self, s):
        return urllib.parse.quote(str(s), safe='~')

File: ruby-sdk/semantria/authrequest.rb
# encoding: utf-8
require 'uri'
require 'base64'
require 'cgi' 
require 'digest/md5'
require 'openssl'
require 'digest/sha1'
require 'net/http'
require 'net/https'

OAuthVersion = "1.0"
OAuthParameterPrefix = "oauth_"
OAuthConsumerKeyKey = "oauth_consumer_key"
OAuthVersionKey = "oauth_version"
OAuthSignatureMethodKey = "oauth_signature_method"
OAuthSignatureKey = "oauth_signature"
OAuthTimestampKey = "oauth_timestamp"
OAuthNonceKey = "oauth_nonce"

class AuthRequest  
  # Create a new instance
  def initialize(consumerKey, consumerSecret, applicationName)
    @consumerKey = consumerKey
    @consumerSecret = consumerSecret
	@applicationName = applicationName
  end

  def authWebRequest(method, url, postData = nil)
    nonce = generateNonce()
    timestamp = generateTimestamp()
    query = generateQuery(method, url, timestamp, nonce)
    authheader = generateAuthHeader(query, timestamp, nonce)
    headers = {"Authorization" => authheader}
    
    if (method == "POST")
      headers["Content-type"] = "application/x-www-form-urlencoded"
    end
	headers["x-api-version"] = "2"
	headers["x-app-name"] = @applicationName
    
    uri = URI.parse(query)
    conn = Net::HTTP.new(uri.host, 443)
    conn.use_ssl = true
    conn.verify_mode = OpenSSL::SSL::VERIFY_NONE

    path = uri.request_uri #'%s?%s' % [qpath, qquery]
    request = getRequest(method, path, headers, postData)
    response = conn.request(request)
 
    result = { "status"=>response.code.to_i, "reason"=>response.message, "data"=>response.body }
    return result
  end
  
  private
  # create the http request object for a given http_method and path
  def getRequest(method, path, headers, postData = nil)
    case method
    when "POST"
      request = Net::HTTP::Post.new(path, headers)
    when "PUT"
      request = Net::HTTP::Put.new(path, headers)
    when "GET"
      request = Net::HTTP::Get.new(path, headers)
    when "DELETE"
      request =  Net::HTTP::Delete.new(path, headers)
    else
      raise ArgumentError, "Don't know how to handle method: :#{method}"
    end
    
    unless postData.nil?
      request.body = postData
      request["Content-Length"] = request.body.length.to_s
    end
    
    request
  end 
  
  def generateQuery(method, url, timestamp, nonce)
    uri = URI.parse(url)
    np = getNormalizedParameters(timestamp, nonce)

    if uri.query
      uri.query = '%s&%s' % [uri.query, np]
    else
      uri.query = '%s' % np
    end
  
    return uri.to_s        
  end

  def generateAuthHeader(query, timestamp, nonce)
    md5cs = getMD5Hash(@consumerSecret)
    escquery = escape(query)
    hash = getSHA1(md5cs, escquery)
    hash = escape(hash)
   
    items = Hash.new()
    items["OAuth realm"] = ""
    items[OAuthVersionKey] = "%s" % OAuthVersion
    items[OAuthTimestampKey] = "%s" % timestamp
    items[OAuthNonceKey] = "%s" % nonce
    items[OAuthSignatureMethodKey] = "HMAC-SHA1"
    items[OAuthConsumerKeyKey] = "%s" % @consumerKey
    items[OAuthSignatureKey] = "%s" % hash
 
    params = []
    for key in items.keys.sort 
       params.push('%s="%s"' % [key, items[key]])
    end
    
    return params.join(',')
  end

  def getNormalizedParameters(timestamp, nonce)
    items = Hash.new()
    items[OAuthVersionKey] = OAuthVersion
    items[OAuthTimestampKey] = timestamp
    items[OAuthNonceKey] = nonce
    items[OAuthSignatureMethodKey] = "HMAC-SHA1"
    items[OAuthConsumerKeyKey] = @consumerKey
    
    params = []
    for key in items.keys.sort 
       params.push('%s=%s' % [key, items[key]])
    end 
        
    np = params.join('&')
    # Encode signature parameters per Oauth Core 1.0 protocol
    # Spaces must be encoded with "%20" instead of "+"
    return np.gsub('+', '%20').gsub('%7E', '~')
  end

  def getMD5Hash(str)
    md5hash = Digest::MD5.hexdigest(str) 
  end
 
  def getSHA1(md5cs, query)
    digest = OpenSSL::Digest::Digest.new('sha1') 
    # our composite signing key now has the token secret after the ampersand
    sha1res = OpenSSL::HMAC.digest(digest, md5cs, query)
    sha1sig = Base64.encode64(sha1res).chomp.gsub(/\n/, "")
  end

  def generateTimestamp()
    Time.now.to_i.to_s
  end

  def generateNonce(length = 20)
    rand(10 ** length).to_s.rjust(length,'0')
  end
  
  def escape(s)
      CGI::escape(s)
  end   
end  